### PR TITLE
[SHA3] Reference spec and cross-spec property tests

### PR DESCRIPTION
**Pull request by abentkamp**
This PR adds the hacspec-style SHA3 reference spec crate (specs/sha3) together with its own test vectors, and wires up cross-spec proptests that check the libcrux-iot implementation against the reference spec.

### CODE REVIEW

**Comment by jschneider-bensch** [comment]
Would it be possible to add the hacspec-style implementation as a git dependency on https://github.com/cryspen/libcrux/tree/sha3-spec-upstream?

EDIT: And eventually just on the main branch, when the spec gets merged there.

**Comment by abentkamp** [comment]
Unfortunately, the specs are slightly different, most notably, the state layout is different (i and j are flipped). Let me see if Claude can fix that and fix the Lean proofs that will break as a consequence.

**Comment by abentkamp** [comment]
I managed to move the specs into libcrux (see also https://github.com/cryspen/libcrux/pull/1472) and to update them to use the main branch of hax-evit.

**Comment by abentkamp** [comment]
Note that one tiny change to libcrux-iot/sha3/src/lib.rs‎ was necessary to make the extraction work with the latest `hax-evit`: I excluded the `Debug` annotation under `cfg(hax_backend_lean)`.

**Comment by robinhundt** [comment]
@abentkamp It seems like currently the cross spec tests are not passing. I suspect it is due to a difference in the indexing in the spec and in the implementation. Can you resolve the test failures?

File: libcrux-iot/sha3/src/keccak.rs [hunk 1]
```diff
@@ -2731,3 +2731,441 @@ impl RotateLeft for U32 {
...
+                ^ state[5 * x + 2]
+                ^ state[5 * x + 3]
+                ^ state[5 * x + 4]
+        })
+    }
+
+    /// Compute the spec's `D` array: `D[x] = C[x-1] ⊕ rot(C[x+1], 1)`.
+    fn spec_d(c: &[u64; 5]) -> [u64; 5] {
+        core::array::from_fn(|x| c[(x + 4) % 5] ^ c[(x + 1) % 5].rotate_left(1))
+    }
```
Comment (robinhundt) [inline libcrux-iot/sha3/src/keccak.rs:2764]: question: Would it make sense to add these to the spec in mainline libcrux?

**Comment by abentkamp** [comment]
Okay, I had fixed those tests, but accidentally put it into the other PR. Fixed now!

File: libcrux-iot/sha3/src/keccak.rs [hunk 2]
```diff
@@ -2731,3 +2731,438 @@ impl RotateLeft for U32 {
...
+                let spec_state = core::array::from_fn(|_| rng.gen::<u64>());
+                let block_u8: [u8; 200] = core::array::from_fn(|_| rng.gen());
+                let block_secret: [libcrux_secrets::U8; 200] =
+                    core::array::from_fn(|i| block_u8[i].classify());
+
+                let spec_out =
+                    hacspec_sha3::sponge::absorb_block(spec_state, &block_u8[..RATE], RATE);
+
+                let mut s = state_from_spec(spec_state);
+                absorb_block::<RATE>(&mut s, &block_secret, 0);
```
Comment (robinhundt) [inline libcrux-iot/sha3/src/keccak.rs]: **suggestion**: This only tests absorb_block with a start offset of 0. It might be good to also test that it is correct if a larger offset is passed.

File: libcrux-iot/sha3/src/keccak.rs [hunk 3]
```diff
@@ -2731,3 +2731,438 @@ impl RotateLeft for U32 {
...
+                    let msg_u8: [u8; 200] = core::array::from_fn(|_| rng.gen());
+                    let msg_secret: [libcrux_secrets::U8; 200] =
+                        core::array::from_fn(|i| msg_u8[i].classify());
+
+                    let spec_out = hacspec_sha3::sponge::absorb_final(
+                        spec_state, &msg_u8, 0, len, RATE, DELIM,
+                    );
+
+                    let mut s = state_from_spec(spec_state);
+                    absorb_final::<RATE, DELIM>(&mut s, &msg_secret, 0, len);
```
Comment (robinhundt) [inline libcrux-iot/sha3/src/keccak.rs]: **suggestion**: This also only tests with a start offset of 0. Also testing with a larger offset would be more robust.

File: libcrux-iot/sha3/src/keccak.rs [hunk 4]
```diff
@@ -2731,3 +2731,438 @@ impl RotateLeft for U32 {
...
+        run::<168>(&mut rng);
+    }
+
+    /// `absorb_final::<RATE, DELIM>` matches `Spec::absorb_final` for partial blocks.
+    #[test]
+    fn absorb_final_matches() {
+        let mut rng = StdRng::seed_from_u64(0xF1AA);
+        fn run<const RATE: usize, const DELIM: u8>(rng: &mut StdRng) {
+            for &len in &[0usize, 1, RATE / 2, RATE - 1] {
+                for _ in 0..2 {
```
Comment (robinhundt) [inline libcrux-iot/sha3/src/keccak.rs]: question: Why is this done twice?

If the intention is to run the test with different randomly generated states, it would be nice if it were consistent with the later used iteration number 4.

File: libcrux-iot/sha3/src/keccak.rs [hunk 5]
```diff
@@ -2731,3 +2731,438 @@ impl RotateLeft for U32 {
...
+        shake_check::<136, 64>(hacspec_sha3::shake256::<64>);
+    }
+    #[test]
+    fn shake256_matches_spec_200() {
+        shake_check::<136, 200>(hacspec_sha3::shake256::<200>);
+    }
+    #[test]
+    fn shake256_matches_spec_512() {
+        shake_check::<136, 512>(hacspec_sha3::shake256::<512>);
+    }
```
Comment (robinhundt) [inline libcrux-iot/sha3/src/keccak.rs:3166]: nit: Can we also test exact multiples of the RATE here? I think 1x and 2x should be sufficient.

File: libcrux-iot/sha3/src/state.rs [hunk 6]
```diff
@@ -196,3 +196,144 @@ impl FromLeBytes<4> for U32 {
...
+
+    /// `load_block` corresponds to `Spec::xor_block_into_state` followed by no permutation.
+    #[test]
+    fn load_block_matches_xor_block_into_state() {
+        let mut rng = StdRng::seed_from_u64(0xBEEF);
+
+        fn run<const RATE: usize>(rng: &mut StdRng) {
+            let initial: [u64; 25] = core::array::from_fn(|_| rng.gen());
+            // load_block reads RATE bytes starting at offset 0.
+            let block_u8: [u8; 200] = core::array::from_fn(|_| rng.gen());
```
Comment (robinhundt) [inline libcrux-iot/sha3/src/state.rs]: question: Is the intention here to allocate a full width array or would a RATE sized one also be possible/more clear?

File: libcrux-iot/sha3/src/state.rs [hunk 7]
```diff
@@ -196,3 +196,144 @@ impl FromLeBytes<4> for U32 {
...
+        let mut rng = StdRng::seed_from_u64(0xFACE);
+
+        fn run<const RATE: usize>(rng: &mut StdRng) {
+            let spec_state: [u64; 25] = core::array::from_fn(|_| rng.gen());
+            let impl_state = state_from_spec(spec_state);
+
+            let spec_out: [u8; RATE] =
+                hacspec_sha3::sponge::squeeze_state::<RATE>(&spec_state, [0u8; RATE], 0, RATE);
+
+            let mut out_secret = [0u8.classify(); 200];
```
Comment (robinhundt) [inline libcrux-iot/sha3/src/state.rs]: question: same as above

File: libcrux-iot/sha3/Cargo.toml [hunk 8]
```diff
@@ -36,14 +36,19 @@ full-unroll = []
...
+    "libcrux-secrets/check-secret-independence",
+    "libcrux-traits/check-secret-independence",
+]
 
 [dev-dependencies]
 criterion = "0.5.1"
 hex = "0.4.3"
 rand = "0.8.5"
-cavp.workspace = true #{ version = "0.0.2-beta.2", path = "../cavp" }
+cavp.workspace = true                                                              #{ version = "0.0.2-beta.2", path = "../cavp" }
```
Comment (robinhundt) [inline libcrux-iot/sha3/Cargo.toml]: nit: cavp is actually unused by now, I created #166 to remove it.

File: libcrux-iot/sha3/src/keccak.rs [hunk 9]
```diff
@@ -2731,3 +2731,438 @@ impl RotateLeft for U32 {
...
+    use rand::rngs::StdRng;
+    use rand::{Rng, SeedableRng};
+
+    /// Recover the underlying u64 from a `Lane2U32` stored in bit-interleaved
+    /// form: deinterleave, then combine the two 32-bit halves.
+    fn lane_interleaved_to_u64(l: &crate::lane::Lane2U32) -> u64 {
+        let d = l.deinterleave();
+        let arr = d.0.declassify();
+        (arr[0] as u64) | ((arr[1] as u64) << 32)
+    }
```
Comment (robinhundt) [inline libcrux-iot/sha3/src/keccak.rs]: nit: Duplicated with the body of `state_to_spec`. Could be moved to state, made `pub(crate)` and use here and `state_to_spec`.

File: libcrux-iot/sha3/src/keccak.rs [hunk 10]
```diff
@@ -2731,3 +2731,438 @@ impl RotateLeft for U32 {
...
+    // ---------------------------------------------------------------------
+
+    /// `absorb_block::<RATE>` corresponds to `Spec::absorb_block`.
+    #[test]
+    fn absorb_block_matches() {
+        let mut rng = StdRng::seed_from_u64(0xABBA);
+        fn run<const RATE: usize>(rng: &mut StdRng) {
+            for _ in 0..4 {
+                let spec_state = core::array::from_fn(|_| rng.gen::<u64>());
+                let block_u8: [u8; 200] = core::array::from_fn(|_| rng.gen());
```
Comment (robinhundt) [inline libcrux-iot/sha3/src/keccak.rs]: question: Is the intention to allocate a full width array here or would a RATE sized one also work?

File: libcrux-iot/sha3/src/keccak.rs (hunk 5 above)
Comment (robinhundt) [reply libcrux-iot/sha3/src/keccak.rs:3166]: Ah, I see that this is done in the proptests. Not needed here then.

File: libcrux-iot/sha3/src/keccak.rs [hunk 11]
```diff
@@ -2731,3 +2731,438 @@ impl RotateLeft for U32 {
...
+    #[test]
+    fn shake256_matches_spec_200() {
+        shake_check::<136, 200>(hacspec_sha3::shake256::<200>);
+    }
+    #[test]
+    fn shake256_matches_spec_512() {
+        shake_check::<136, 512>(hacspec_sha3::shake256::<512>);
+    }
+
+    extern crate alloc;
```
Comment (robinhundt) [inline libcrux-iot/sha3/src/keccak.rs]: nit: This would be more visible at the top of the module.

**Review by robinhundt — CHANGES_REQUESTED** [review]
Thanks for the PR and moving the specs to mainline libcrux!

Most of the comments are Nits or question. The one thing that I think should be added is testing that the `absorb_block` function is also correct with non-zero offsets, as that is actually used in the keccak function.

Regarding the `for _ in 2` (or 4) loops, I assume this is to test with different randomly generated states for a couple of iterations? For that, 2 seems a bit small and it would be nice if it were consistent.

EDIT: I realized after requesting changes that the `absorb_block` and `absorb_final` functions are tested with different offsets indirectly in the proptests. Adding unit tests for these functions would be nice to have, but not blocking.

File: libcrux-iot/sha3/src/keccak.rs (hunk 10 above)
Comment (robinhundt) [reply libcrux-iot/sha3/src/keccak.rs]: Actually, if the test would also test non-zero offsets, it makes sense to allocate inputs larger than RATE.

File: libcrux-iot/sha3/src/keccak.rs (hunk 2 above)
Comment (robinhundt) [reply libcrux-iot/sha3/src/keccak.rs]: I noticed that the proptests do indirectly test absorb_block with offsets as there larger messages are hashed, but I still think it would be valuable to specifically test that here in a unit test as well.

**Comment by robinhundt** [comment]
I've resolved my comments and from my side this is good to go @jschneider-bensch. The only thing left is potentially moving the two spec functions needed here into the mainline spec.

File: libcrux-iot/sha3/src/keccak.rs (hunk 1 above)
Comment (jschneider-bensch) [reply libcrux-iot/sha3/src/keccak.rs:2764]: Indeed, they could be part of the spec at some point, but I don't think it's critical.

File: libcrux-iot/sha3/Cargo.toml [hunk 12]
```diff
@@ -36,13 +36,19 @@ full-unroll = []
...
+    "libcrux-traits/check-secret-independence",
+]
 
 [dev-dependencies]
 criterion = "0.5.1"
 hex = "0.4.3"
 rand = "0.8.5"
+cavp.workspace = true #{ version = "0.0.2-beta.2", path = "../cavp" }
 pretty_env_logger = "0.5.0"
+hacspec_sha3 = { git = "https://github.com/cryspen/libcrux.git", rev = "b313344" }
```
Comment (jschneider-bensch) [inline libcrux-iot/sha3/Cargo.toml]: We should drop the `rev` field, once https://github.com/cryspen/libcrux/pull/1399 is merged.

Comment (robinhundt) [reply libcrux-iot/sha3/Cargo.toml]: I've updated the rev to the merge commit where the spec is merged.
I'd prefer to keep the rev, so the sha3 cross spec tests/lean proof are not automatically broken should we make changes to the spec.

Comment (jschneider-bensch) [reply libcrux-iot/sha3/Cargo.toml]: True!
If we change the spec in a breaking way that would be something we'd want to react to quickly, though, right? Keeping it pinned might mean we don't notice for a while that we're working against an old spec.

Comment (robinhundt) [reply libcrux-iot/sha3/Cargo.toml]: Yes, we might not notice for some time. But I'd argue that is better than having the repo in broken state. Because we don't have a lock file, a change to the spec would not only immediately break the current and future versions (or commits) of libcrux-iot, but also past commits with an updated lockfile.  
But even if we don't pin the rev here, we might not notice breakage of the cross spec tests, if we don't run the tests (e.g. after the project concludes and there are no new CI runs for some time). Pinning the rev here ensures that the tests will execute for a fresh checkout of libcrux-iot.

Comment (robinhundt) [reply libcrux-iot/sha3/Cargo.toml]: Ideally, the spec would also be released and versioned on crates.io and we'd have dependabot updates for the spec here. Then we can't overlook it, but can update the cross spec tests here when convenient instead of being forced to.

Comment (jschneider-bensch) [reply libcrux-iot/sha3/Cargo.toml]: > Ideally, the spec would also be released and versioned on crates.io and we'd have dependabot updates for the spec here. Then we can't overlook it, but can update the cross spec tests here when convenient instead of being forced to.

Yes, I agree that having a released (more importantly, versioned, IMO) spec would be nice.
The points you raise about breaking the repo retroactively make a lot of sense. Then we should keep the rev here until we've come to a re-evualtion re. lock files and versioning of specs.
